engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -617,14 +733,35 @@ private static boolean atomOverridesEquals(Atom atom) {
   @Fallback
   @TruffleBoundary
   boolean equalsGeneric(Object left, Object right,
-      @CachedLibrary(limit = "5") InteropLibrary interop) {
+      @CachedLibrary(limit = "5") InteropLibrary interop,
+      @CachedLibrary(limit = "5") TypesLibrary typesLib) {
       return left == right
           || interop.isIdentical(left, right, interop)
-          || left.equals(right);
+          || left.equals(right)
+          || (isNullOrNothing(left, typesLib, interop) && isNullOrNothing(right, typesLib, interop));
+  }
+
+  private boolean isNullOrNothing(Object object, TypesLibrary typesLib, InteropLibrary interop) {
+    if (typesLib.hasType(object)) {
+      return typesLib.getType(object) == EnsoContext.get(this).getNothing();
```

Comment:
I don't think you need `TypeLibrary` here. Enough to `object == EnsoContext.get(this).getNothing()`. `Nothing` is a special Enso value where `Nothing.type == Nothing`.